libs/libskynet/src/seed.ts

```diff
@@ -0,0 +1,202 @@
+}
+
+// validSeedPhrase checks whether the provided seed phrase is valid, returning
+// an error if not. If the seed phrase is valid, the full seed will be returned
+// as a Uint8Array.
+function validSeedPhrase(seedPhrase: string): [Uint8Array, string | null] {
+  // Create a helper function to make the below code more readable.
+  const prefix = function (s: string): string {
+    return s.slice(0, DICTIONARY_UNIQUE_PREFIX);
+  };
+
+  // Pull the seed into its respective parts.
+  const seedWordsAndChecksum = seedPhrase.split(" ");
+  const seedWords = seedWordsAndChecksum.slice(0, SEED_ENTROPY_WORDS);
+  const checksumOne = seedWordsAndChecksum[SEED_ENTROPY_WORDS];
+  const checksumTwo = seedWordsAndChecksum[SEED_ENTROPY_WORDS + 1];
+
+  // Convert the seedWords to a seed.
+  const [seed, err1] = seedWordsToSeed(seedWords);
+  if (err1 !== null) {
+    return [new Uint8Array(0), addContextToErr(err1, "unable to parse seed phrase")];
+  }
+
+  const [checksumOneVerify, checksumTwoVerify, err2] = seedToChecksumWords(seed);
+  if (err2 !== null) {
+    return [new Uint8Array(0), addContextToErr(err2, "could not compute checksum words")];
+  }
+  if (prefix(checksumOne) !== prefix(checksumOneVerify)) {
+    return [new Uint8Array(0), "first checksum word is invalid"];
+  }
+  if (prefix(checksumTwo) !== prefix(checksumTwoVerify)) {
+    return [new Uint8Array(0), "second checksum word is invalid"];
+  }
+  return [seed, null];
+}
+
+// seedWordsToSeed will convert a provided seed phrase to to a Uint8Array that
+// represents the cryptographic seed in bytes.
+function seedWordsToSeed(seedWords: string[]): [Uint8Array, string | null] {
+  // Input checking.
+  if (seedWords.length !== SEED_ENTROPY_WORDS) {
+    return [
+      new Uint8Array(0),
+      `Seed words should have length ${SEED_ENTROPY_WORDS} but has length ${seedWords.length}`,
+    ];
+  }
+
+  // We are getting 16 bytes of entropy.
+  const bytes = new Uint8Array(SEED_BYTES);
+  let curByte = 0;
+  let curBit = 0;
+  for (let i = 0; i < SEED_ENTROPY_WORDS; i++) {
+    // Determine which number corresponds to the next word.
+    let word = -1;
+    for (let j = 0; j < dictionary.length; j++) {
+      if (seedWords[i].slice(0, DICTIONARY_UNIQUE_PREFIX) === dictionary[j].slice(0, DICTIONARY_UNIQUE_PREFIX)) {
+        word = j;
+        break;
+      }
+    }
+    if (word === -1) {
+      return [new Uint8Array(0), `word '${seedWords[i]}' at index ${i} not found in dictionary`];
+    }
+    let wordBits = 10;
+    if (i === SEED_ENTROPY_WORDS - 1) {
+      wordBits = 8;
+    }
+
+    // Iterate over the bits of the 10- or 8-bit word.
+    for (let j = 0; j < wordBits; j++) {
+      const bitSet = (word & (1 << (wordBits - j - 1))) > 0;
+
+      if (bitSet) {
+        bytes[curByte] |= 1 << (8 - curBit - 1);
+      }
+
+      curBit += 1;
+      if (curBit >= 8) {
+        // Current byte has 8 bits, go to the next byte.
+        curByte += 1;
+        curBit = 0;
+      }
+    }
+  }
```

Comment:
all lines with byte shifts and increments require a comment on what's happening there